File: 2024/3/regex.py
import sys
import re

if __name__ == "__main__":
    mul_do_dont = re.compile(r"mul\((\d+),(\d+)\)|do\(\)|don't\(\)")
    data = sys.stdin.read()

    solution1 = 0
    solution2 = 0
    do = True
    for m in re.finditer(mul_do_dont, data):
        if m.group(0) == "do()":
            do = True
        elif m.group(0) == "don't()":
            do = False
        else:
            v = int(m.group(1)) * int(m.group(2))
            solution1 += v
            if do:
                solution2 += v
    print(solution1)
    print(solution2)

File: 2024/3/regex.go
//go:build go_regex

package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strconv"
)

var (
	mulDoDontPattern = regexp.MustCompile(`mul\((\d+),(\d+)\)|do\(\)|don't\(\)`)
)

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	var data string
	do := true
	solution1 := 0
	solution2 := 0
	for scanner.Scan() {
		data = scanner.Text()

		for _, match := range mulDoDontPattern.FindAllStringSubmatch(data, -1) {
			if match[0] == "do()" {
				do = true
			} else if match[0] == "don't()" {
				do = false
			} else {
				x, _ := strconv.Atoi(match[1])
				y, _ := strconv.Atoi(match[2])
				v := x * y
				solution1 += v
				if do {
					solution2 += v
				}
			}

		}
	}

	fmt.Println(solution1)
	fmt.Println(solution2)
}
